### PR TITLE
Feat: implement Notifier, Notice & Notification (COR-24)

### PR DESCRIPTION
**Pull request by tristanwagner**
# Description

Implement an api to display global notifications.

Fixes #COR-24

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation

# How Has This Been Tested?

Ran some examples in a vue 3 project

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/components/Notification/Notification.module.scss [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+        &_hook_holder {
+            position: relative;
+        }
+
+        &_notice {
+            background: #fff;
+            border-radius: $border-radius-base;
+            box-shadow: 0 3px 6px -4px rgb(0 0 0 / 12%), 0 6px 16px 0 rgb(0 0 0 / 8%), 0 9px 28px 8px rgb(0 0 0 / 5%);
+            line-height: 1.5715;
```
Comment (josermarinr) [inline src/components/Notification/Notification.module.scss:32]: Very specific line-height

File: src/components/Notification/Notification.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+} as Meta;
+
+const Template: StoryFn<NotificationArgsProps> = (args) => {
+
+  return (
+    <div>
+      <button onClick={() => notification.open(args)}>Open notification</button>
+    </div>
+  );
+};
```
Comment (josermarinr) [inline src/components/Notification/Notification.stories.tsx:36]: `const Template: StoryFn<NotificationArgsProps> = (args) => (
    <div>
      <button onClick={() => notification.open(args)}>Open notification</button>
    </div>
  );`

File: src/components/Notification/Notification.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  closable: true,
+  key: '',
+  class: '',
+  style: {},
+  top: '24px',
+  bottom: '24px',
+  getContainer: () => document.body,
+  onClick: () => alert('onClick'),
+  onClose: () => console.log('onClose'),
+};
```
Comment (josermarinr) [inline src/components/Notification/Notification.stories.tsx:54]: bad indentation in this file

File: src/components/Notification/Notification.module.scss [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+$notification-width: 384px;
+$notification-padding-vertical: 16px;
+$notification-padding-horizontal: 24px;
+$notification-padding: $notification-padding-vertical $notification-padding-horizontal;
+$font-size-base: 14px;
+$font-size-lg: $font-size-base + 2px;
+$font-size-sm: $font-size-base - 2px;
+$border-radius-base: 2px;
+$heading-color: fade(#000, 85%);
+
```
Comment (josermarinr) [inline src/components/Notification/Notification.module.scss:12]: bad indentation to stylint

File: src/components/Notification/Notification.module.scss [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        }
+
+        &_notice {
+            background: #fff;
+            border-radius: $border-radius-base;
+            box-shadow: 0 3px 6px -4px rgb(0 0 0 / 12%), 0 6px 16px 0 rgb(0 0 0 / 8%), 0 9px 28px 8px rgb(0 0 0 / 5%);
+            line-height: 1.5715;
+            margin-bottom: $notification-margin-bottom;
+            margin-left: auto;
+            max-width: calc(100vw - $notification-margin-edge * 2);
```
Comment (josermarinr) [inline src/components/Notification/Notification.module.scss:35]: instead 100%

File: src/components/Notification/Notification.tsx [hunk 6]
```diff
@@ -0,0 +1,271 @@
+import classNames from "classnames";
+import {CSSProperties} from "vue";
+import {renderHelper} from "../../util/helpers";
+import CorCheckIcon from "../Icons/CorCheckIcon/CorCheckIcon";
+import Notifier from "../Notifier/Notifier";
+import {NotifierInstance} from "../Notifier/Notifier.types";
+import {ConfigProps, IconType, NotificationApi, NotificationArgsProps, NotificationPlacement} from "./Notification.types";
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:7]: IconType is not used

File: src/components/Notification/Notification.tsx [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+const notificationInstances: { [key: string]: NotifierInstance } = {};
+
+// default values
+let defaultDuration = 5;
+let defaultTop = '24px';
+let defaultBottom = '24px';
+let defaultPrefixCls = 'cor_notification';
+let defaultPlacement: NotificationPlacement = 'topRight';
+let defaultGetContainer = () => document.body;
+let defaultCloseIcon: any  = null;
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:20]: let defaultCloseIcon: VueNode | (() => VueNode) = null;

File: src/components/Notification/Notification.tsx [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+      }
+      break;
+    case 'bottomRight':
+      style = {
+        right: '0px',
+        top: 'auto',
+        bottom,
+      }
+      break;
+  }
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:121]: default case?

File: src/components/Notification/Notification.tsx [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        }
+
+        let iconNode = null;
+        if (icon) {
+          iconNode = () => <span class={`${prefixCls}_icon`}>{renderHelper(icon)}</span>
+        } else if (type) {
+          const Icon = typeToIcon[type];
+          // TODO:
+          // iconNode = () => <Icon class={`${prefixCls}_icon ${prefixCls}_icon_${type}]`} />
+        }
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:222]: it not used

File: src/components/Notification/Notification.tsx [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+            {btn ? <span class={`${prefixCls}_btn`}>{renderHelper(btn)}</span> : null}
+          </div>
+        );
+      }
+    })
+  })
+}
+
+// api we are exposing to the users
+const api: any = {
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:243]: use typing

Comment (josermarinr) [reply src/components/Notification/Notification.tsx:243]: or delete any

File: src/components/Notification/Notification.tsx [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+// const iconTypes: IconType[] = ['success', 'info', 'warning', 'error'];
+// iconTypes.forEach(type => {
+//   api[type] = (args: NotificationArgsProps) =>
+//     api.open({
+//       ...args,
+//       type,
+//     });
+// });
+
+// api.warn = api.warning;
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:269]: Delete

File: src/components/Notification/Notification.tsx [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+              )
+            }
+          </span>
+        );
+      },
+      maxCount,
+      transitionName: 'fade',
+      hasTransitionName: true,
+    },
+    (notificationInstance: any) => {
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:178]: NotifierInstance

File: src/components/Notifier/Notifier.tsx [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+      const notice: NoticeContent & { key: Key; userPassKey?: Key } = {
+        ...originNotice,
+        key,
+      };
+      const { maxCount } = props;
+      const noticeIndex = notices.value.map(v => v.notice.key).indexOf(key);
+      const updatedNotices = notices.value.concat();
+
+      if (noticeIndex !== -1) {
+        updatedNotices.splice(noticeIndex, 1, { notice, holderCallback } as any);
```
Comment (josermarinr) [inline src/components/Notifier/Notifier.tsx:61]: Type instantiation is excessively deep and possibly infinite

File: src/components/Notifier/Notifier.tsx [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+          ...notice.props,
+          key,
+          noticeKey: userPassKey || key,
+          updateMark,
+          onClose: (noticeKey: Key) => {
+            remove(noticeKey);
+            notice.onClose?.();
+          },
+          onClick: notice.onClick,
+        };
```
Comment (josermarinr) [inline src/components/Notifier/Notifier.tsx:111]: Type instantiation is excessively deep and possibly infinite.

File: src/components/Notifier/Notifier.tsx [hunk 15]
```diff
@@ -0,0 +1,237 @@
+import {
```
Comment (josermarinr) [inline src/components/Notifier/Notifier.tsx:1]: Same than Notification.tsx I think is better practices if we cout in many file and scoop responsibility of function by file

File: src/components/Notifier/Notifier.tsx (hunk 13 above)
Comment (josermarinr) [reply src/components/Notifier/Notifier.tsx:61]: https://en.wikipedia.org/wiki/Modular_programming <---- check it

File: src/components/Notifier/Notice.tsx [hunk 16]
```diff
@@ -0,0 +1,114 @@
+import { computed, defineComponent, onMounted, watch, onUnmounted } from 'vue';
+import type { CSSProperties } from 'vue';
+import classNames, { Argument } from 'classnames';
+import { NoticeProps } from './Notice.types';
+
+export default defineComponent<NoticeProps>({
+  name: 'Notice',
```
Comment (josermarinr) [inline src/components/Notifier/Notice.tsx:7]: Component name "Notice" should always be multi-word. => Cor-Notice maybe?

File: src/components/Notifier/Notice.tsx [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+  ] as any,
+  setup(props, { attrs, slots }) {
+
+    let closeTimer: any;
+    const duration = computed(() => (props.duration === undefined ? 3 : props.duration));
+
+    const startCloseTimer = () => {
+      if (duration.value) {
+        closeTimer = setTimeout(() => {
+          close();
```
Comment (josermarinr) [inline src/components/Notifier/Notice.tsx:28]: 'close' was used before it was defined

File: src/components/Notifier/Notice.tsx [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+      const node = (
+        <div
+          class={classNames(componentClass, className as Argument, {
+            [`${componentClass}_closable`]: closable,
+          })}
+          style={style as CSSProperties}
+          onMouseenter={clearCloseTimer}
+          onMouseleave={startCloseTimer}
+          onClick={onClick as any}
```
Comment (josermarinr) [inline src/components/Notifier/Notice.tsx:99]: (payload: MouseEvent) => void

Comment (josermarinr) [reply src/components/Notifier/Notice.tsx:99]: typing

File: src/components/Notification/Notification.module.scss [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+$notification-padding-vertical: 16px;
+$notification-padding-horizontal: 24px;
+$notification-padding: $notification-padding-vertical $notification-padding-horizontal;
+$font-size-base: 14px;
+$font-size-lg: $font-size-base + 2px;
+$font-size-sm: $font-size-base - 2px;
+$border-radius-base: 2px;
+$heading-color: fade(#000, 85%);
+
+:global {
```
Comment (josermarinr) [inline src/components/Notification/Notification.module.scss:13]: you should delete global
use scss file not a module.scss instead

File: src/components/Notification/Notification.tsx [hunk 20]
```diff
@@ -0,0 +1,271 @@
+import classNames from "classnames";
+import {CSSProperties} from "vue";
+import {renderHelper} from "../../util/helpers";
+import CorCheckIcon from "../Icons/CorCheckIcon/CorCheckIcon";
+import Notifier from "../Notifier/Notifier";
+import {NotifierInstance} from "../Notifier/Notifier.types";
+import {ConfigProps, IconType, NotificationApi, NotificationArgsProps, NotificationPlacement} from "./Notification.types";
+import './Notification.module.scss'
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:8]: use import './Notification.scss'

File: src/components/Notifier/Notifier.tsx (hunk 14 above)
Comment (tristanwagner) [reply src/components/Notifier/Notifier.tsx:111]: yea typescript has a bad type system

File: src/components/Notification/Notification.stories.tsx (hunk 3 above)
Comment (tristanwagner) [reply src/components/Notification/Notification.stories.tsx:53]: done

File: src/components/Notification/Notification.module.scss (hunk 4 above)
Comment (tristanwagner) [reply src/components/Notification/Notification.module.scss]: done

File: src/components/Notification/Notification.module.scss (hunk 5 above)
Comment (tristanwagner) [reply src/components/Notification/Notification.module.scss]: done

File: src/components/Notifier/Notice.tsx (hunk 16 above)
Comment (tristanwagner) [reply src/components/Notifier/Notice.tsx]: i named it v-notice and also v-notifier good for you ?

File: src/components/Notifier/Notice.tsx (hunk 17 above)
Comment (tristanwagner) [reply src/components/Notifier/Notice.tsx]: done

File: src/components/Notifier/Notice.tsx (hunk 18 above)
Comment (tristanwagner) [reply src/components/Notifier/Notice.tsx]: done

File: src/components/Notification/Notification.module.scss (hunk 19 above)
Comment (tristanwagner) [reply src/components/Notification/Notification.module.scss]: done

File: src/components/Notification/Notification.tsx (hunk 20 above)
Comment (tristanwagner) [reply src/components/Notification/Notification.tsx]: done

File: src/components/Notification/Notification.stories.tsx (hunk 2 above)
Comment (tristanwagner) [reply src/components/Notification/Notification.stories.tsx]: done

File: src/components/Notification/Notification.tsx [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+                }
+
+                let iconNode = null;
+                if (icon) {
+                    iconNode = () => <span class={`${prefixCls}_icon`}>{renderHelper(icon)}</span>;
+                } else if (type) {
+                    // TODO:
+                    // const Icon = typeToIcon[type];
+                    // iconNode = () => <Icon class={`${prefixCls}_icon ${prefixCls}_icon_${type}]`} />
+                }
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:214]: line not used

File: src/components/Notification/Notification.tsx [hunk 22]
```diff
@@ -0,0 +1,263 @@
...
+// const iconTypes: IconType[] = ['success', 'info', 'warning', 'error'];
+// iconTypes.forEach(type => {
+//   api[type] = (args: NotificationArgsProps) =>
+//     api.open({
+//       ...args,
+//       type,
+//     });
+// });
+
+// api.warn = api.warning;
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:261]: idem dead code

File: src/components/Notification/Notification.types.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    closable?: boolean;
+    render?: VueNode | ((config: NotificationArgsProps) => VueNode);
+}
+
+// TODO: implement support for success, error, info, warning
+export interface NotificationInstance {
+    // success(args: NotificationArgsProps): void;
+    // error(args: NotificationArgsProps): void;
+    // info(args: NotificationArgsProps): void;
+    // warning(args: NotificationArgsProps): void;
```
Comment (josermarinr) [inline src/components/Notification/Notification.types.ts:55]: code not used

File: src/components/Notification/Notification.tsx [hunk 24]
```diff
@@ -0,0 +1,263 @@
...
+    if (options.rtl !== undefined) {
+        rtl = options.rtl;
+    }
+    if (options.maxCount !== undefined) {
+        maxCount = options.maxCount;
+    }
+    if (options.closable !== undefined) {
+        defaultClosable = options.closable;
+    }
+}
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:64]: should be an other file

File: src/components/Notification/Notification.tsx [hunk 25]
```diff
@@ -0,0 +1,263 @@
...
+        );
+    },
+    config: setNotificationConfig,
+    destroy() {
+        Object.keys(notificationInstances).forEach((cacheKey) => {
+            notificationInstances[cacheKey].destroy();
+            delete notificationInstances[cacheKey];
+        });
+    },
+};
```
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:249]: api or Notification

File: src/components/Notification/Notification.tsx (hunk 21 above)
Comment (josermarinr) [inline src/components/Notification/Notification.tsx:214]: should be other file

File: src/components/Notification/Notification.tsx (hunk 24 above)
Comment (tristanwagner) [reply src/components/Notification/Notification.tsx:64]: if its in another file I can't override default values

File: src/components/Notifier/Notifier.tsx (hunk 14 above)
Comment (josermarinr) [reply src/components/Notifier/Notifier.tsx]: 🤡 nice

File: src/components/Notifier/Notice.tsx (hunk 16 above)
Comment (josermarinr) [reply src/components/Notifier/Notice.tsx]: Cor for cornellius

**Review by josermarinr — APPROVED** [review]
nice job

File: src/components/Notifier/Notice.tsx (hunk 16 above)
Comment (tristanwagner) [reply src/components/Notifier/Notice.tsx]: ok I renamed to 'cor-notice' and 'cor'-notifier'

**Pull request merged**